build/resolver.go

```diff
@@ -0,0 +1,365 @@
+package build
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/containerd/containerd/platforms"
+	"github.com/docker/buildx/builder"
+	"github.com/docker/buildx/driver"
+	"github.com/docker/buildx/util/progress"
+	"github.com/moby/buildkit/client"
+	gateway "github.com/moby/buildkit/frontend/gateway/client"
+	"github.com/moby/buildkit/util/flightcontrol"
+	"github.com/moby/buildkit/util/tracing"
+	specs "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+	"go.opentelemetry.io/otel/trace"
+	"golang.org/x/sync/errgroup"
+)
+
+type resolvedNode struct {
+	resolver    *nodeResolver
+	driverIndex int
+	platforms   []specs.Platform
+}
+
+func (dp resolvedNode) Node() builder.Node {
+	return dp.resolver.nodes[dp.driverIndex]
+}
+
+func (dp resolvedNode) Client() *client.Client {
+	clients, err := dp.resolver.boot(context.TODO(), []int{dp.driverIndex}, nil)
+	if err != nil {
+		panic(err)
+	}
+	return clients[0]
+}
+
+func (dp resolvedNode) Opts() gateway.BuildOpts {
+	opts, err := dp.resolver.opts(context.TODO(), []int{dp.driverIndex}, nil)
+	if err != nil {
+		panic(err)
+	}
+	return opts[0]
+}
+
+type matcher struct {
+	platforms.Matcher
+	perfect bool
+}
+
+type nodeResolver struct {
+	nodes    []builder.Node
+	clients  flightcontrol.Group[*client.Client]
+	opt      flightcontrol.Group[gateway.BuildOpts]
+	matchers []func(specs.Platform) matcher
+}
+
+func newDriverResolver(nodes []builder.Node) *nodeResolver {
+	// create a collection of ordered matchers for each platform
+	var matcherCache sync.Map
+	lookup := func(p specs.Platform) []matcher {
+		p = platforms.Normalize(p)
+		pk := platforms.Format(p)
+		if matchers, ok := matcherCache.Load(pk); ok {
+			return matchers.([]matcher)
+		}
+
+		// We want to match against the strictest possible platform first,
+		// descending to the least strict. If the match is not perfect, we
+		// want to boot all the nodes to get the full list of available
+		// platforms and retry all the perfect matchers before trying the
+		// inperfect ones.
+		var matchers []matcher
+		matchers = append(matchers, matcher{platforms.OnlyStrict(p), true})
+		matchers = append(matchers, matcher{platforms.Only(p), true})
+		matchers = append(matchers, matcher{newArchMatcherStrict(p), false})
+		matchers = append(matchers, matcher{newArchMatcher(p), false})
```

Comment:
I think we could add a specific case where `arm` arch would match `arm64`.

Reply:
Ah damn forgot it's already the case nevermind https://github.com/docker/buildx/blob/master/vendor/github.com/containerd/containerd/platforms/compare.go#L70-L82